packages/ui/slide-over.tsx

```diff
@@ -0,0 +1,162 @@
+'use client';
+
+import {
+  Fragment,
+  useState,
+  ReactNode,
+  createContext,
+  useContext,
+  Dispatch,
+} from 'react';
+import {Dialog, Transition} from '@headlessui/react';
+import {XMarkIcon} from '@heroicons/react/24/outline';
+import classNames from 'classnames';
+
+interface SlideOverProps {
+  children: ReactNode;
+  variant?: 'text' | 'gallery';
+}
+
+interface SlideOverContextType {
+  isOpen: boolean;
+  setIsOpen: Dispatch<React.SetStateAction<boolean>>;
+  variant: SlideOverProps['variant'];
+}
+
+const SlideOverContext = createContext<SlideOverContextType>({
+  isOpen: false,
+  setIsOpen: () => {},
+  variant: 'text',
+});
+
+function SlideOver({children, variant = 'text'}: SlideOverProps) {
+  const [isOpen, setIsOpen] = useState(false);
+
+  const context = {
+    isOpen,
+    setIsOpen,
+    variant,
+  };
+  return (
+    <SlideOverContext.Provider value={context}>
+      {children}
+    </SlideOverContext.Provider>
+  );
+}
+
+interface SlideOverDialogProps {
+  children: ReactNode;
+}
+
+function SlideOverDialog({children}: SlideOverDialogProps) {
+  const {isOpen, setIsOpen, variant} = useContext(SlideOverContext);
+
+  const panelClassName = classNames('pointer-events-auto w-screen', {
+    'max-w-md bg-white': variant === 'text',
+    'max-w-4xl bg-neutral-600': variant === 'gallery',
+  });
+
+  return (
+    <Transition.Root show={isOpen} as={Fragment}>
+      <Dialog
+        as="div"
+        className="relative z-10"
+        onClose={() => setIsOpen(false)}
+      >
+        <div className="fixed inset-0" />
+
+        <div className="fixed inset-0 overflow-hidden">
+          <div className="absolute inset-0 overflow-hidden">
+            <div className="pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10">
+              <Transition.Child
+                as={Fragment}
+                enter="transform transition ease-in-out duration-500 sm:duration-700"
+                enterFrom="translate-x-full"
+                enterTo="translate-x-0"
+                leave="transform transition ease-in-out duration-500 sm:duration-700"
+                leaveFrom="translate-x-0"
+                leaveTo="translate-x-full"
+              >
+                <Dialog.Panel className={panelClassName}>
+                  <div className="flex h-full flex-col shadow-xl">
+                    {children}
+                  </div>
+                </Dialog.Panel>
+              </Transition.Child>
+            </div>
+          </div>
+        </div>
+      </Dialog>
+    </Transition.Root>
+  );
+}
+
+interface OpenButtonProps {
+  children: ReactNode;
+  className?: string;
```

Comment:
To be sure: is `className` allowed to be `undefined`?

Reply:
Yes. If you want a button with no styling you can leave it empty